src/python/grpcio/grpc/_cython/_cygrpc/aio/iomgr/iomgr.pyx.pxi

```diff
@@ -0,0 +1,185 @@
+# Copyright 2019 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from cpython cimport Py_INCREF, Py_DECREF
+
+from libc cimport string
+
+cdef grpc_socket_vtable asyncio_socket_vtable
+cdef grpc_custom_resolver_vtable asyncio_resolver_vtable
+cdef grpc_custom_timer_vtable asyncio_timer_vtable
+cdef grpc_custom_poller_vtable asyncio_pollset_vtable
+
+
+cdef grpc_error* asyncio_socket_init(
+        grpc_custom_socket* grpc_socket,
+        int domain) with gil:
+    socket = _AsyncioSocket.create(grpc_socket)
+    Py_INCREF(socket)
+    grpc_socket.impl = <void*>socket
+    return <grpc_error*>0
+
+
+cdef void asyncio_socket_destroy(grpc_custom_socket* grpc_socket) with gil:
+    Py_DECREF(<_AsyncioSocket>socket.impl)
+
+
+cdef void asyncio_socket_connect(
+        grpc_custom_socket* grpc_socket,
+        const grpc_sockaddr* addr,
+        size_t addr_len,
+        grpc_custom_connect_callback connect_cb) with gil:
+
+    host, port = sockaddr_to_tuple(addr, addr_len)
+    socket = <_AsyncioSocket>grpc_socket.impl
+    socket.connect(host, port, connect_cb)
+
+
+cdef void asyncio_socket_close(
+        grpc_custom_socket* grpc_socket,
+        grpc_custom_close_callback close_cb) with gil:
+    socket = (<_AsyncioSocket>grpc_socket.impl)
+    if socket.is_connected():
+        socket.writer.close()
+    close_cb(grpc_socket)
+
+
+cdef void asyncio_socket_shutdown(grpc_custom_socket* grpc_socket) with gil:
+    raise NotImplemented()
```

Comment:
Can you implement this function, it is required for internal build. Here is the version I wrote for tests to pass.
```Cython
cdef void asyncio_socket_shutdown(grpc_custom_socket* grpc_socket) with gil:
    socket = (<_AsyncioSocket>grpc_socket.impl)
    if socket.is_connected():
        socket._writer.close()
```

Reply:
There is a slightly different implementation of this, already in https://github.com/Skyscanner/grpc/pull/8
It basically calls ``socket.close()`` as the ``socket.is_connected()`` verification is done internally.
Mentioning it in case it serves as reference, and also to watch out for conflicts.